### PR TITLE
driver import reorganization: clean docs build

### PR DESCRIPTION
**Pull request by loriab**
## Todos

- [x] Collect enough extra docstrings and tweak enough references to functions so that all the links in the docs resolve.
  - [x] Ended up needing mild to moderate docs renovation through whole QCA stack
  - [x] Now can turn on Sphinx warnings-to-errors and nitpicky so that docs remain healthy
- [x] Added new doc extension dependency, https://github.com/mansenfranzen/autodoc_pydantic, so that Pydantic models (as introduced in #2575) now get superior, rather than inferior, autodoc compared to classes.
- [x] Add in a few more `__all__`s to driver files.

## Checklist
- ~Tests added for any new features~
- [x] Full tests run

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

File: doc/sphinxman/source/add_tests.rst [hunk 1]
```diff
@@ -289,10 +289,12 @@ Plain Old Data
 .. comment compare_arrays covered by compare_values
 
 .. autofunction:: qcelemental.testing.compare_values
+   :noindex:
 
 .. autofunction:: psi4.compare_integers(expected, computed [, label, *, **kwargs])
 
 .. autofunction:: qcelemental.testing.compare
+   :noindex:
```
Comment (JonathonMisiewicz) [inline doc/sphinxman/source/add_tests.rst:297]: When is `noindex` needed?

File: doc/sphinxman/source/dftd3.rst [hunk 2]
```diff
@@ -373,6 +373,8 @@ any user-specified dispersion parameters.
 
 .. autofunction:: qcdb.Molecule.run_dftd3
 
+.. autofunction:: qcdb.Molecule.run_dftd4
```
Comment (JonathonMisiewicz) [inline doc/sphinxman/source/dftd3.rst:376]: When is `autofunction` needed?

File: doc/sphinxman/source/glossary_psivariables.rst [hunk 3]
```diff
@@ -1477,6 +1477,11 @@ PSI Variables by Alpha
 
    A induction-classified SAPT term energy [Eh] implemented for SAPT2+3.
 
+.. psivar:: SAPT EXCH-IND30(S^INF) ENERGY
```
Comment (JonathonMisiewicz) [inline doc/sphinxman/source/glossary_psivariables.rst:1480]: Glad to see that nothing I did created problems here for you.

File: doc/sphinxman/source/nbody.rst [hunk 4]
```diff
@@ -79,6 +79,12 @@ The nbody function computes counterpoise-corrected (CP), non-CP (noCP), and Vali
     # Returns the nocp energy as its first in the list
     energy('CCSD(T)', bsse_type=['nocp', 'cp', 'vmfc'], max_nbody=3)
 
-.. autoclass:: psi4.driver.driver_nbody.ManyBodyComputer
+API
+---
+
+.. autoclass:: psi4.driver.driver_nbody.BsseEnum
    :members:
+   :undoc-members:
```
Comment (JonathonMisiewicz) [inline doc/sphinxman/source/nbody.rst:87]: What do `:members:` and `:undoc-members:` mean?

File: doc/sphinxman/source/psi4api.rst [hunk 5]
```diff
@@ -39,4 +39,16 @@
 .. automodapi:: psi4.driver
 
 .. automodapi:: psi4.driver.p4util
+   :skip:ConvergenceError
```
Comment (JonathonMisiewicz) [inline doc/sphinxman/source/psi4api.rst:42]: I know I'm asking this a lot, but what does `:skip:` do?

**Review by JonathonMisiewicz — COMMENTED** [review]
The PR looks good to me, but I don't understand a few changes. I may just need to be pointed to a Sphinx syntax manual.

I'll be happy to approve this once you check the "ready for review" box. Otherwise, I'll assume there are more commits to come.

File: doc/sphinxman/source/dftd3.rst (hunk 2 above)
Comment (loriab) [reply doc/sphinxman/source/dftd3.rst:376]: Basically, whenever in narrative docs users would benefit from seeing the (hopefully better maintained) docstring from the code, throw in the autofunction or autoclass. Also, if some narrative or docstring text mentions a fn or class with 

```
:class:`psi4.core.set_memory`
:func:`~psi4.driver.energy`  # tilde means link looks like "energy", not "psi4.driver.energy"
```

then Sphinx wants to link to it. If it hasn't already been covered by an automodapi, then you need to add the autofunction somewhere as a link target. (Miscellaneous end up in `external_apis.rst`.)

File: doc/sphinxman/source/glossary_psivariables.rst (hunk 3 above)
Comment (loriab) [reply doc/sphinxman/source/glossary_psivariables.rst:1480]: All's well. The `-n` nitpicky option shows up missing glossary entries easily. Unfortunately docs weren't in a shape to use it during your excited state qcvars pass.

**Review by JonathonMisiewicz — APPROVED** [review]

**Review by maxscheurer — APPROVED** [review]

**Review by jeffschriber — APPROVED** [review]

**Pull request merged**